### PR TITLE
[toward #5102, refuse ancient servers] login: Equalize user-facing errors across both api.getUserSettings calls

### PR DESCRIPTION
**Pull request by chrisbobbe**
And make them a bit more descriptive.

(EDIT: I've updated the "After" list below for the revision I sent earlier today: https://github.com/zulip/zulip-mobile/pull/5591#issuecomment-1347530790)

-----

## Before:

- On submitting the realm input, an error alert with title "Cannot connect to server." and no body
- On picking a logged-out account in the account-pick screen, an error alert with title "Failed to connect to server: {realm}" and no body.

-----

## After:

In both places, an error alert with title "Could not connect", and one of these as the body:

- On `ApiError`: "The server at {realm} said:\n\n{message}" (where `realm` is a realm URL and `message` is, [e.g.](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/.2Fserver_settings.3A.20.60realm_name.60.2C.20etc.2E/near/1332900), "Subdomain required")

- On `NetworkError`: "Could not connect to {realm}. Please check your network connection and try again." (where `realm` is a realm URL)

- On `MalformedResponseError` with `httpStatus` 404: "The server at {realm} does not seem to be a Zulip server."

- On `Server5xxError`: "The server at {realm} encountered an error."

- Otherwise: "Failed to connect to server: {realm}" (where `realm` is a realm URL)

Here are screen recordings of example error alerts:

### CODE REVIEW

**Comment by chrisbobbe** [comment]
FYI @alya for the UX change.

**Comment by alya** [comment]
Seems reasonable to me... Can we give any helpful advice in the "Failed to connect to server" case? What might be wrong, and what can the user try?

**Review by gnprice — COMMENTED** [review]
Thanks! Generally this looks good; comments below.

File: src/account/AccountPickScreen.js [hunk 1]
```diff
@@ -48,9 +49,31 @@ export default function AccountPickScreen(props: Props): Node {
...
+              'Could not connect to {realm}. Please check your network connection and try again.',
+              { realm: realm.toString() },
+            );
+          } else {
+            msg = _('Failed to connect to server: {realm}', { realm: realm.toString() });
+          }
+
+          // If changing the alert text, see if we should make a similar change
+          // for the other api.getServerSettings call.
+          // TODO(?): Deduplicate
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply src/account/AccountPickScreen.js:75]: Makes sense. Any preference where the new function `fetchServerSettings` gets defined?

**Comment by chrisbobbe** [comment]
> Can we give any helpful advice in the "Failed to connect to server" case? What might be wrong, and what can the user try?

That one is meant as a fallback and should only show after we've checked for known common things that could go wrong, like an error with the network connection, and we've come up empty so we can't really suggest things to try.

A few things come to mind to make that check smarter, though, so we don't end up using the fallback as much:
- We should probably have a case for 5xx errors, so we can say the server is busted and you should bug an admin or try again later.
- Probably a somewhat common failure mode is that you put in a URL that's not a Zulip server at all, like `recurse.com` instead of `recurse.zulipchat.com`. Greg, should we try to think of some test for whether the server is a Zulip server?

-----

Oh, while I'm here, what do we think of the following error alert for #5102, refusing to connect to ancient servers:

Title: "Could not connect" (same as the ones in this PR)
Body: "{realm} is running Zulip Server {version}, which is unsupported."
"Learn more" button: https://zulip.readthedocs.io/en/stable/overview/release-lifecycle.html#compatibility-and-upgrading

File: src/account/AccountPickScreen.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply src/account/AccountPickScreen.js:75]: Yeah, I think I'd leave it outside of `src/api/`.

Maybe `fetchActions.js`? Fits well enough there, I think, though not perfectly.

**Comment by gnprice** [comment]
> * Probably a somewhat common failure mode is that you put in a URL for not a Zulip server at all, like `recurse.com` instead of `recurse.zulipchat.com`. Greg, if we're thinking of some test for whether the server is a Zulip server, does anything come to mind? Maybe if we get a 404, we just say it's not a Zulip server? (It could still be a misconfigured one, but there's no way to be foolproof with this.)

Yeah, I think if we try to fetch the server-settings route and get back a 404 — so it's not that the server is unreachable or down, but rather the server is informing us it has no such route — then it's fair to conclude that it's not a Zulip server.



> Oh, while I'm here, what do we think of the following error alert for #5102, refusing to connect to ancient servers (will be a separate PR):

Seems reasonable to me!

One thought that brings to mind is that it would be good to have a Help Center page we can link to instead of that developer-facing page. The Help Center page could potentially link in turn to that page, but first it would give a focused answer to the questions a user might have, like:
* why am I being shown this;
* what do I need to ask my IT department (in general, my server admin) to do.

Probably we'd want users to get pointed at that Help Center page to send it to their server admins when asking them for help, too ­— so it should have a clear action item for them with a link to upgrade instructions.

**Comment by chrisbobbe** [comment]
Makes sense, thanks for the review!

> Help Center page

Just filed zulip/zulip#23842 for this.

**Comment by chrisbobbe** [comment]
Revision pushe.d

**Comment by chrisbobbe** [comment]
Ah, and I've just updated the description with the current list of user-facing messages, and two screen recordings.

@alya FYI.

**Comment by gnprice** [comment]
Thanks! Looks good to me.

I'll leave it open for @alya to comment too.

**Comment by alya** [comment]
Is there a good way to avoid the line break after `https://`? Otherwise looks good to me!

**Comment by chrisbobbe** [comment]
> Is there a good way to avoid the line break after `https://`? Otherwise looks good to me!

Hmm, yeah I agree that's a bad place for the line to break.

Unfortunately we can't easily control how these iOS alerts are styled; it's a native component rather than something managed in React. @gnprice, what do you think?

**Comment by gnprice** [comment]
> Unfortunately we can't easily control how these iOS alerts are styled; it's a native component rather than something managed in React. @gnprice, what do you think?

Yeah, it doesn't seem like we have a good way to control that. So we'll just let it be.

The same thing is true in the underlying iOS SDK, too — this isn't a case where there's functionality underneath which RN just isn't exposing to us. It's using UIAlertController, of which the `message` property is just a string:
https://developer.apple.com/documentation/uikit/uialertcontroller/1620106-message

**Pull request merged**